src/System/Ctags/Internal.hs

```diff
@@ -13,25 +20,47 @@ import System.Ctags.IO
 import System.Ctags.Types
 import qualified System.Directory as D
 
-tagsContent :: MonadIO m => m (Either TagSearchOutcome T.Text)
-tagsContent = readTagsFile =<< findTagsFile
+type AppConfig = MonadReader CtagsSearchConfig
+
+newtype App a = App
+    { runApp :: ReaderT CtagsSearchConfig (ExceptT TagSearchOutcome IO) a
+    } deriving ( Functor
+               , Applicative
+               , Monad
+               , AppConfig
+               , MonadIO
+               , MonadError TagSearchOutcome
+               )
+
+data CtagsSearchConfig = CtagsSearchConfig
+    { tagFileName :: String
+    , tagFileDirectories :: [String]
+    }
+
+defaultCtagsSearchConfig :: CtagsSearchConfig
+defaultCtagsSearchConfig = CtagsSearchConfig "tags" [".git", "tmp", "."]
+
+tagsContent ::
+       MonadIO m => CtagsSearchConfig -> m (Either TagSearchOutcome T.Text)
+tagsContent config = liftIO $ runExceptT (runReaderT (runApp run) config)
+
+run :: App T.Text
+run = readTagsFile =<< findTagsFile
 
 stdinContent :: MonadIO m => m T.Text
 stdinContent = lenientUtf8Decode <$> getContentsLazy
 
-findTagsFile :: MonadIO m => m (Maybe FilePath)
-findTagsFile = findFile possibleTagsFileDirectories "tags"
-  where
-    findFile dirs = liftIO . D.findFile dirs
+findTagsFile :: (MonadIO m, AppConfig m) => m (Maybe FilePath)
+findTagsFile = do
+    directories <- asks tagFileDirectories
+    filename <- asks tagFileName
+    liftIO $ D.findFile directories filename
 
-readTagsFile :: MonadIO m => Maybe String -> m (Either TagSearchOutcome T.Text)
-readTagsFile Nothing =
-    return $ Left $ TagsFileNotFound possibleTagsFileDirectories
+readTagsFile :: Maybe String -> App T.Text
+readTagsFile Nothing = throwError . TagsFileNotFound =<< asks tagFileDirectories
```

Comment:
If you move the `ExceptT` into `findTagsFile`, this won't need `ReaderT` anymore and you can remove the pattern match.

Reply:
Ah, interesting - like this? 23c8018926056efb32db94096bc765482f1880ba